### PR TITLE
[webgpu] Use i32 as coords type in WGSL

### PR DESCRIPTION
**Pull request by axinging**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5628)

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @haoyunfeix @xhcao @gyagp  PTAL

A comment, we can use 
1. getXXXAtOutCoordsByGlobalId(vec3<i32>(globalId), index) to get data.
2. Or we can define an extra globalIdInt:
```
let globalIdInt = vec3<i32>globalId;
```
Then use getXXXAtOutCoordsByGlobalId(index)  to get data.

As to the 1, it will save a vec3. As to the 2, it will save some code changes. Current  1 is applied.

File: tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts [hunk 1]
```diff
@@ -245,48 +245,48 @@ const SHADER_PREFIX = `
   }
 
   // Checks whether coordinates lie within the bounds of the shape.
-  fn coordsInBounds4D(coord : vec4<u32>, shape : vec4<u32>) -> bool {
+  fn coordsInBounds4D(coord : vec4<i32>, shape : vec4<i32>) -> bool {
     return all(coord < shape);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts:249]: Should your previous changes be reverted #5559? Use `return all(coord >= vec4<u32>(0u)) && all(coord < shape);`. And similar for other places.

File: tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts [hunk 2]
```diff
@@ -245,48 +245,48 @@ const SHADER_PREFIX = `
...
-    return u32(dot(vec4<f32>(coords), vec4<f32>(
+  fn getFlatIndex4D(coords : vec4<i32>, shape : vec4<i32>) -> i32 {
+    return i32(dot(vec4<f32>(coords), vec4<f32>(
         f32(shape.y) * f32(shape.z) * f32(shape.w), f32(shape.z) * f32(shape.w), f32(shape.w), 1.0)));
   }
 
   // Only used when the y/z dimension of workgroup size is 1.
-  fn getGlobalIndex(globalId : vec3<u32>, localId : vec3<u32>) -> u32 {
-    if (uniforms.dispatchSize.y == 1u && uniforms.dispatchSize.z == 1u) {
+  fn getGlobalIndex(globalId : vec3<i32>, localId : vec3<i32>) -> i32 {
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts:279]: How about changing it like below. And if it's the only place that `dispatchSize` is used, you can keep it as `u32` for convenience.
And `numWorkGroups` sounds like a better name compared with `dispatchSize`.
```
fn getGlobalIndex(globalId : vec3<u32>, localId : vec3<u32>) -> i32
{
   XXX
   XXX
   return i32(XXX);
}
```

File: tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts [hunk 3]
```diff
@@ -487,26 +483,26 @@ export class MatMulPackedVec4Program implements WebGPUProgram {
 
     const userCode = `
       ${activationSnippet}
-      fn mm_readA(row : u32, col : u32,  globalId : vec3<u32>) -> vec4<f32> {
+      fn mm_readA(row : i32, col : i32,  globalId : vec3<i32>) -> vec4<f32> {
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts:486]: I prefer you keep the globalId's type as `vec3<u32>`. And just change `let batch = i32(globalId.z);` which is simpler.

**Review by qjia7 — COMMENTED** [review]
I vote keeping `globalId`, `localId`'s type as function parameter types. And do the corresponding cast as needed. And for others, we all unify to use `i32` instead of `u32` expect for `dispatchSize`. DispatchSize or `numWorkGroups` is similarly like `globalId`, `localId` which is u32 instead i32.

File: tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts [hunk 4]
```diff
@@ -150,11 +150,11 @@ export class Conv2DDerInputMMProgram implements WebGPUProgram {
     if (xC < 0.0 || xC >= f32(uniforms.outBackprop[2]) || fract(xC) > 0.0) {
       return 0.0;
     }
-    let coord = vec4<u32>(
+    let coord = vec4<i32>(
         batch,
-        u32(xR),
-        u32(xC),
-        col % u32(uniforms.outBackprop[3]));
+        i32(xR),
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts:155]: Cast to i32 not needed?

File: tfjs-backend-webgpu/src/kernels/im2col_webgpu.ts [hunk 5]
```diff
@@ -95,28 +96,28 @@ export class Im2ColProgram implements WebGPUProgram {
     const userCode = `
     ${getWorkGroupSizeStringWgsl()}
     fn main([[builtin(global_invocation_id)]] globalId : vec3<u32>) {
-      let index = globalId.x;
+      let index = i32(globalId.x);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/im2col_webgpu.ts:99]: It seems that you should use `getGlobleIndex` here.

**Review by qjia7 — APPROVED** [review]
Looks better, thanks. LGTM with two nits.

File: tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts (hunk 4 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts:155]: Here xR is f32.

File: tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts [hunk 6]
```diff
@@ -140,8 +140,8 @@ export class Conv2DDerInputMMProgram implements WebGPUProgram {
     let outRow = row / uniforms.outShape[2];
     let outCol = row % uniforms.outShape[2];
 
-    let WRow = i32(col) / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
-    let WCol = i32(col) / uniforms.outBackprop[3] % uniforms.filterDims[1];
+    let WRow = col / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
+    let WCol = col / uniforms.outBackprop[3] % uniforms.filterDims[1];
     let xR = (f32(outRow) - f32(uniforms.pads[0]) + f32(WRow)) / f32(uniforms.stride[0]);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts:145]: Can it be changed to `let xR = f32(outRow - uniforms.pads[0] + WRow) / f32(uniforms.stride[0]);`? And same to `xc`.

File: tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts (hunk 4 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/conv_backprop_mm_webgpu.ts:155]: Yes, I missed it. Thanks.

**Review by qjia7 — APPROVED** [review]

**Review by lina128 — APPROVED** [review]
Thank you for the refactor, looks great!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5628#-:-MjjZ0Ap2YOE4_UBSe0b:b-c1uyp4)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @axinging and @pyu10055)

**Pull request merged**